src/common/progress_bar.cpp

```diff
@@ -0,0 +1,68 @@
+#include "duckdb/common/progress_bar.hpp"
+
+namespace duckdb {
+void ProgressBar::PrintProgress(int percentage) {
+	int lpad = (int)(percentage / 100.0 * pbwidth);
+	int rpad = pbwidth - lpad;
+	printf("\r%3d%% [%.*s%*s]", percentage, lpad, pbstr.c_str(), rpad, "");
+	fflush(stdout);
+}
+
+void ProgressBar::ProgressBarThread() {
+	std::this_thread::sleep_for(std::chrono::milliseconds(show_progress_after));
+	while (!StopRequested()) {
+		int percentage = executor->GetPipelinesProgress(supported);
+		if (supported && percentage > 0) {
+			PrintProgress(percentage);
+		}
+
+		std::this_thread::sleep_for(std::chrono::milliseconds(time_update_bar));
+	}
+}
+
+void ProgressBar::ProgressBarThreadTest() {
+	std::this_thread::sleep_for(std::chrono::milliseconds(show_progress_after));
+	while (!StopRequested()) {
+		int percentage = executor->GetPipelinesProgress(supported);
+		bool acceptable_percentage = cur_percentage <= percentage && percentage <= 100;
+		valid_percentage = valid_percentage && acceptable_percentage;
```

Comment:
Does the testing code need to be integrated into the progress bar directly? Can't you just always set `cur_percentage`, and read that in the tests?